packages/@webex/plugin-meetings/src/breakouts/index.ts

```diff
@@ -0,0 +1,206 @@
+  },
+
+  derived: {
+    isInMainSession: {
+      deps: ['sessionType'],
+      /**
+       * Returns true if the user is in the main session
+       * @returns {boolean}
+       */
+      fn() {
+        return this.sessionType === BREAKOUTS.SESSION_TYPES.MAIN;
+      }
+    },
+  },
+
+  /**
+   * initialize for the breakouts
+   * @returns {void}
+   */
+  initialize() {
+    this.listenTo(this, 'change:status', () => {
+      if (this.status === BREAKOUTS.STATUS.CLOSING) {
+        this.trigger(BREAKOUTS.EVENTS.BREAKOUTS_CLOSING);
+      }
+    });
+    this.debouncedQueryRosters = debounce(this.queryRosters, 10, {
+      leading: true,
+      trailing: false
+    })
+    this.listenTo(this.breakouts, 'add', () => {
+      this.debouncedQueryRosters();
+    });
+    this.listenToBroadcastMessages();
+    this.listenToBreakoutRosters();
+  },
+
+  /**
+   * Calls this to clean up listeners
+   */
+  cleanUp() {
+    this.stopListening()
+  },
+
+  /**
+   * Update the current locus url of the meeting
+   */
+  locusUrlUpdate(locusUrl) {
+    this.set('locusUrl', locusUrl);
+  },
+
+  /**
+   * The initial roster lists need to be quried because you don't
+   * get a breakout.roster event when you join the meeting
+   */
+  queryRosters() {
+    this.webex.request(
+      {uri: `${this.url}/roster`, qs: {locusUrl: btoa(this.locusUrl)}
+    }).then((result) => {
+      const {body: {rosters}} = result;
+      
+      rosters.forEach(({locus}) => {
+        this.handleRosterUpdate(locus);
+      })
+
+      this.trigger(BREAKOUTS.EVENTS.MEMBERS_UPDATE);
+    })
+  },
+
+  handleRosterUpdate(locus) {
+    const sessionId = locus.controls?.breakout?.sessionId;
+
+    const session = this.breakouts.get(sessionId);
+
+    if (!session) {
+      return;
+    }
+
+    session.parseRoster(locus);
+  },
+
+  /**
+   * Sets up listener for broadcast messages sent to the breakout session
+   */
+  listenToBroadcastMessages() {
+    this.listenTo(this.webex.internal.llm, 'event:breakout.message', (event) => {
+      const {data: {senderUserId, sentTime, message}} = event;
+
+      this.trigger(BREAKOUTS.EVENTS.MESSAGE, {
+        senderUserId,
+        sentTime,
+        message,
+        // FIXME: This is only the current sessionId
+        // We'd need to check that the dataChannelUrl is still the same
+        // to guarantee that this message was sent to this session
+        sessionId: this.currentBreakoutSession.sessionId,
+      });
+    })
+  },
+
+  /**
+   * Sets up a listener for roster messags from mecury
+   */
+  listenToBreakoutRosters() {
+    this.listenTo(this.webex.internal.mercury, 'event:breakout.roster', (event) => {
+      this.handleRosterUpdate(event.data.locus);
+      this.trigger(BREAKOUTS.EVENTS.MEMBERS_UPDATE);
+    })
+  },
+
+  /**
+   * Updates the information about the current breakout
+   * @param {Object} params
+   * @returns {void}
+   */
+  updateBreakout(params) {
+    this.set(params);
```

Comment:
why is this needed?

Reply:
You mean why not call .set directly? In the rest of the sdk we typically don't access built in methods of ampersand state directly. After all, ampersand state is basically an implementation choice and so exposing its API to something outside of the unit might make it more brittle. Typically we define methods like updateWhatever and then other parts of the code don't have to know that set exists, or how it works. Adding a separate method also gives us room to sanitize input in the future

Reply:
I was referring to why calling this.set() since it's called below again with more parameters. I guess it would be nice to describe what is `params` or how does the object look like.

Reply:
yes, perhaps we could define a type somehow to show what exactly gets set on the object. Ampersand state ignores the params that are not defined anyway.